### PR TITLE
Create-CustomTraefikImage

### PR DESCRIPTION
**Pull request by sirhc101**
Create a custom Traefik docker image, based on the best generic image version according to your Windows version.
This is useful when using Windows Server 2022 for example, where no prebuilt image is provided.

### CODE REVIEW

**Comment by freddydk** [comment]
Thanks for creating a function, which is on my todo-list:-)

**Comment by sirhc101** [comment]
I've added the release note entry and updated the manifest files.
Additionally I moved the function into 'traefik' subfolder, so both traefik related functions are placed in the same directory

**Comment by freddydk** [comment]
Sorry for leaving this on the floor for almost a year.
Is this still relevant? then please pull the latest changes from containerhelper and resolve the conflicts.
Also remove the change in .psd1 as this is autogenerated.

**Comment by sirhc101** [comment]
> Sorry for leaving this on the floor for almost a year. Is this still relevant? then please pull the latest changes from containerhelper and resolve the conflicts. Also remove the change in .psd1 as this is autogenerated.

Yes, this is still relevant. I've updated the files

**Comment by freddydk** [comment]
If you can fix the three issues pointed out by the review comments?

**Comment by sirhc101** [comment]
Sorry for asking, but which three issues in the review comments do you mean?

File: ContainerHandling/traefik/Create-CustomTraefikImage.ps1 [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+                  Write-Host "Commandline: docker $arguments" -ForegroundColor Red
+              }
+          }
+          return $result
+      }
+
+
+      try {
+          $bestGenericImage = Get-BestGenericImageName
+          $servercoreVersion = $bestGenericImage.Split(':')[1]
```
Comment (freddydk) [inline ContainerHandling/traefik/Create-CustomTraefikImage.ps1:90]: BestGenericImageName might be 

`mcr.microsoft.com/businesscentral:10.0.20348.1668-dev
`

So you would need $bestGenericImage.Split(':')[1].Split('-')[0]

**Review by freddydk — CHANGES_REQUESTED** [review]

**Comment by freddydk** [comment]
> Sorry for asking, but which three issues in the review comments do you mean?

My bad - I hadn't submitted my review - sorry.

**Review by freddydk — CHANGES_REQUESTED** [review]

File: ContainerHandling/traefik/Create-CustomTraefikImage.ps1 [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+          $bestGenericImage = Get-BestGenericImageName
+          $servercoreVersion = $bestGenericImage.Split(':')[1].Split('-')[0]
+          $serverCoreImage = "mcr.microsoft.com/windows/servercore:$serverCoreVersion"
+
+          Write-Host "Pulling $serverCoreImage (this might take some time)"
+          if (!(DockerDo -imageName $serverCoreImage -command 'pull'))  {
+              throw "Error pulling image"
+          }
+          
+          if ((-not $traefikVersion) -or ($traefikVersion -eq "")) {
```
[the remark this answers is not in the document]
Comment (sirhc101) [reply ContainerHandling/traefik/Create-CustomTraefikImage.ps1]: Of course, you're right! Changed this

**Comment by freddydk** [comment]
Thanks, will merge and do some tests

**Comment by freddydk** [comment]
Did some testing and some modifications:
- You were right that mandatory should not be used, then the default value doesn't matter and you have to specify traefikversion
- We should only update the traefikImage in containerhelper.config.json - there might be other changes to the containerhelper config session, which should not be written
- Function should return the imagename created (including tag)
- Default tag should be traefikVersion instead of latest
- traefik.exe should be downloaded outside the docker build (as a lot of machines won't have internet access inside the docker build - including mine)
- temp folder should not be in c:\ - but instead underneath containerhelperconfig.hostHelperFolder and always be deleted.

I fixed those and will check in a new PR, integrate this in aka.ms/getbc and test more there.

**Comment by sirhc101** [comment]
Many thanks @freddydk 🙏🏼

**Pull request merged**